### ISSUE TITLE
Code review. Part 4

### ISSUE THREAD

**Issue by ablaom** [alan-turing-institute/MLJFlux.jl#31]
Here are some suggestions for improving ImageClassifier:

- [x] Replace the existing `collate` method. The new generic `collate` in `core.jl` should do the trick after one implements appropriate `reformat` and `get` commands. Something like this might work but needs checking (add a test for this case of `collate` to test/core.jl):

```julia
reformat(X, ::Type{<:AbstractVector{<:GrayImage}}) = X
function get(X::AbstractVector{<:AbstractMatrix{<:Gray}}, b)
    ret = Array{Float32}(undef, size(first(X)..., 1, length(b))
    for i in eachindex(b)
        ret[:, :, :, i] = Float32.(X[b[i]])
    end
    return ret
end
```

Note that the existing collate assumes the target is integer-valued, with a value from 0 to 9 (remnant of the MNIST POC, I guess). The target is anything with scitype `<:AbstractArray{<:Finite}`. The existing `reformat` and `get` should deal with this already.

- [x] Assuming we are happy with #30 , do the same here

- [x] Tidy up the update logic, as in src/classifier.jl

- [x] Add tests (currently there are none at all), which could be modelled on the vanilla classifier in test/classifier.jl. To keep test time down, I suggest we construct a synthetic image data set with image size 2 x 2 or similar.

- [ ] Separately retest on MNIST (I think there is notebook for this somewhere?) for final sanity check

- [ ] Generalise code and tests to ColorImage case #4

**Comment by ayush-1506** [comment]
@ablaom Quick question: I'm trying to implement the reformat and get methods for collate, but adding `reformat(X, ::Type{<:AbstractVector{<:GrayImage}}) = X` gives me:
```
┌ Warning: No convention specified. Did you forget to use the `set_convention` function?
└ @ ScientificTypes ~/.julia/packages/ScientificTypes/Gnt2M/src/ScientificTypes.jl:72
```
What's the workaround for this?

**Comment by ablaom** [comment]
Have you loaded `MLJBase`? You need this for all testing. The `scitype` method (and others) have a minimal implementation when you only import `MLJModelInterface`.

**Comment by ablaom** [comment]
To fix the 1.0 fail, move the struct definition near top of image.jl into global scope (ie outside of and before `@testset` block).

```julia
    mutable struct mynn <: MLJFlux.Builder
        kernel1
        kernel2
    end
```
